File: packages/oxlint-plugin-react-doctor/src/plugin/rules/preact/preact-prefer-oninput.ts
import { defineRule } from "../../utils/define-rule.js";
import { findJsxAttribute } from "../../utils/find-jsx-attribute.js";
import { getJsxPropStringValue } from "../../utils/get-jsx-prop-string-value.js";
import type { EsTreeNode } from "../../utils/es-tree-node.js";
import type { EsTreeNodeOfType } from "../../utils/es-tree-node-of-type.js";
import { isNodeOfType } from "../../utils/is-node-of-type.js";
import type { Rule } from "../../utils/rule.js";

const PREFER_ONINPUT_MESSAGE =
  "In Preact core, `onChange` on text-like inputs only fires on blur — use `onInput` for real-time updates. If using `preact/compat`, this is handled automatically.";

// Input types where the native DOM `change` event fires on blur (not on
// every keystroke). Matches the set exempted by preact/compat's
// `onChangeInputType` regex `/fil|che|rad/` — everything NOT matching
// that regex is affected.
const COMPAT_EXEMPT_INPUT_TYPES = new Set(["checkbox", "radio", "file"]);

// Import sources that signal preact/compat is active in this file.
// When compat is in the dependency graph, onChange is automatically
// remapped to onInput for affected elements — no lint needed.
const COMPAT_IMPORT_SOURCES = new Set([
  "preact/compat",
  "react",
  "react-dom",
  "react/jsx-runtime",
  "react/jsx-dev-runtime",
  "react-dom/client",
]);

const fileImportsPreactCompat = (program: EsTreeNodeOfType<"Program">): boolean => {
  for (const statement of program.body) {
    if (!isNodeOfType(statement as EsTreeNode, "ImportDeclaration")) continue;
    const source = (statement as EsTreeNodeOfType<"ImportDeclaration">).source;
    const value =
      source && typeof (source as { value?: unknown }).value === "string"
        ? (source as { value: string }).value
        : null;
    if (!value) continue;
    if (COMPAT_IMPORT_SOURCES.has(value)) return true;
  }
  return false;
};

const isTextLikeInput = (openingElement: EsTreeNodeOfType<"JSXOpeningElement">): boolean => {
  if (!isNodeOfType(openingElement.name, "JSXIdentifier")) return false;
  const tagName = openingElement.name.name;
  if (tagName === "textarea") return true;
  if (tagName !== "input") return false;
  const typeAttribute = findJsxAttribute(openingElement.attributes as EsTreeNode[], "type");
  if (!typeAttribute) return true;
  const typeValue = getJsxPropStringValue(typeAttribute);
  if (typeValue === null) return true;
  return !COMPAT_EXEMPT_INPUT_TYPES.has(typeValue);
};

// In Preact core (without preact/compat), the native DOM `change` event
// on text-like `<input>` and `<textarea>` elements fires only when the
// element loses focus — not on every keystroke. React famously remaps
// `onChange` to the native `input` event for these elements;
// `preact/compat` mirrors that remapping. Pure Preact projects that
// import from `preact` directly must use `onInput` for real-time updates.
//
// This rule only fires when the file does NOT import from `preact/compat`,
// `react`, or `react-dom` — the presence of any compat-layer import
// means the remapping is active and `onChange` works as expected.
export const preactPreferOninput = defineRule<Rule>({
  id: "preact-prefer-oninput",
  requires: ["preact"],
  severity: "warn",
  recommendation:
    "Replace `onChange` with `onInput` on text-like inputs, or use `preact/compat` which remaps `onChange` automatically.",
  create: (context) => {
    let hasCompat = false;

    return {
      Program(node: EsTreeNodeOfType<"Program">) {
        hasCompat = fileImportsPreactCompat(node);
      },
      JSXOpeningElement(node: EsTreeNodeOfType<"JSXOpeningElement">) {
        if (hasCompat) return;
        if (!isTextLikeInput(node)) return;
        const onChangeAttribute = findJsxAttribute(node.attributes as EsTreeNode[], "onChange");
        if (!onChangeAttribute) return;
        context.report({ node: onChangeAttribute, message: PREFER_ONINPUT_MESSAGE });
      },
    };
  },
});

File: native/oxlint/rules/preact-prefer-oninput.rs
use oxc_ast::AstKind;
use oxc_diagnostics::OxcDiagnostic;
use oxc_macros::declare_oxc_lint;

use crate::{AstNode, context::LintContext, rule::Rule};

const MESSAGE: &str = "Your users see no live updates because `onChange` on text inputs in Preact core only fires on blur, so use `onInput` instead. `preact/compat` handles this for you.";
const EXEMPT_INPUT_TYPES: [&str; 3] = ["checkbox", "radio", "file"];

#[derive(Debug, Default, Clone)]
pub struct PreactPreferOninput;

declare_oxc_lint!(
    /// Prefer onInput for text-like controls in Preact core.
    PreactPreferOninput,
    react_doctor_native,
    correctness,
    version = "0.1.0",
    short_description = "Prefer onInput for text-like controls in Preact core.",
);

impl Rule for PreactPreferOninput {
    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
        let AstKind::JSXOpeningElement(opening_element) = node.kind() else {
            return;
        };
        let Some((element_type, _)) = resolve_jsx_element_type(opening_element, ctx) else {
            return;
        };
        if element_type != "textarea"
            && (element_type != "input"
                || find_jsx_attribute(opening_element, "type")
                    .and_then(|attribute| attribute.value.as_ref())
                    .and_then(|value| get_direct_string_literal_attribute_value(value))
                    .is_some_and(|input_type| EXEMPT_INPUT_TYPES.contains(&input_type)))
        {
            return;
        }
        let Some(on_change_attribute) = find_jsx_attribute(opening_element, "onChange") else {
            return;
        };
        ctx.diagnostic(OxcDiagnostic::warn(MESSAGE).with_label(on_change_attribute.span));
    }
}
